src/box/memtx_engine.cc

```diff
@@ -1264,6 +1265,37 @@ struct memtx_join_ctx {
 	struct xstream *stream;
 };
 
+static void
+send_join_meta(struct xstream *stream, const struct vclock *vclock,
+	       const bool send_meta, const struct raft_request *raft_req,
+	       const struct synchro_request *synchro_req)
+{
+	/* Respond to the JOIN request with the current vclock. */
+	struct xrow_header row;
+	/* Encoding replication request uses fiber()->gc region. */
+	RegionGuard region_guard(&fiber()->gc);
+	xrow_encode_vclock(&row, vclock);
+	xstream_write(stream, &row);
+
+	if (send_meta) {
+		/* Mark the beginning of the metadata stream. */
+		xrow_encode_type(&row, IPROTO_JOIN_META);
+		xstream_write(stream, &row);
+
+		xrow_encode_raft(&row, &fiber()->gc, raft_req);
+		xstream_write(stream, &row);
+
+		char body[XROW_BODY_LEN_MAX];
+		xrow_encode_synchro(&row, body, synchro_req);
+		row.replica_id = synchro_req->replica_id;
+		xstream_write(stream, &row);
+
+		/* Mark the end of the metadata stream. */
+		xrow_encode_type(&row, IPROTO_JOIN_SNAPSHOT);
+		xstream_write(stream, &row);
```

Comment:
Then send it conditionally? My main point is that `IPROTO_JOIN_SNAPSHOT` is not the end of meta section. It is beginning of snapshot data. Which means it can't be in this function which is only for sending the meta.

Reply:
> Then send it conditionally?

I see, fixed

Reply:
Thanks for this small detail 🙏!